File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    
    HashMap<String, Integer> phone = new HashMap<>();
    
    public boolean solution(String[] phone_book) {
        boolean answer = true;
        
        for(int i = 0; i < phone_book.length; i++){
            phone.put(phone_book[i], 1);
        }

        for(int i = 0; i < phone_book.length; i++){
            String temp = "";
            for(int j = 0; j < phone_book[i].length() - 1; j++){
                temp += phone_book[i].charAt(j);
                if(phone.containsKey(temp)){
                    return false;
                }
            }
        }
        return true;
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;
map<string, bool> dict;

bool solution(vector<string> phone_book) {
    bool answer = true;
    
    for(int i = 0; i <phone_book.size(); i++){
        dict[phone_book[i]] = true;
    }
    
    for(int i = 0; i < phone_book.size(); i++){
        string temp = "";
        for(int j = 0; j < phone_book[i].length(); j++){
            temp += phone_book[i][j];
            if(dict[temp] && temp != phone_book[i])
                answer = false;
        }
        if(!answer)
            break;
    }
    return answer;
}